sb_vision/native/cv3d.cpp

```diff
@@ -0,0 +1,100 @@
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <stdbool.h>
+#include <string>
+#include <vector>
+
+extern "C" {
+    int solve_pnp(
+        const double object_points[],
+        const double image_points[],
+        const double camera_matrix[],
+        const double dist_coeffs[],
+        double rvec[],
+        double tvec[]
+    );
+}
+
+#include "opencv2/opencv.hpp"
+
+void print_array(const int rows, const int cols, const double values[]) {
+    printf("[\n");
+    for (int i=0; i<rows; i++) {
+        printf("  [");
+        for (int j=0; j<cols; j++) {
+            int idx = (i * cols) + j;
+            printf("%f", values[idx]);
+            if (j+1 < cols) {
+                printf(", ");
+            }
+        }
+        printf("],\n");
+    }
+    printf("]\n");
+}
+
+void print_mat(cv::Mat& mat) {
+    printf("(%d x %d = %d) ", mat.rows, mat.cols, mat.total());
+    printf("[\n");
+    for (int i=0; i<mat.rows; i++) {
+        printf("  [");
+        for (int j=0; j<mat.cols; j++) {
+            double * val = (double *)mat.ptr(i, j);
+            printf("%f", *val);
+            if (j+1 < mat.cols) {
+                printf(", ");
+            }
+        }
+        printf("],\n");
+    }
+    printf("]\n");
+}
+
+#define COPY_INTO(mat, data) \
+    memcpy(mat.ptr(), data, mat.rows * mat.cols * sizeof(double));
+
+
+int solve_pnp(
+    const double object_points[],
+    const double image_points[],
+    const double camera_matrix[],
+    const double dist_coeffs[],
+    double rvec[],
+    double tvec[]
+) {
+    // Copy our input data into OpenCV's Mat wrapper
+    cv::Mat object_points_mat(4, 3, CV_64FC1);
```

Comment:
I've seen other implementations of this use `cv::DataType<double>::type` instead of `CV_64FC1`, I'm not sure if the former is nicer though.

Reply:
Interesting, I think that that's probably nicer as it's clearer what the constant means.

Reply:
5d32563 does this; thanks! (I'd not realised that that existed)